cpp/src/arrow/compute/kernels/vector_selection_take_internal.cc

```diff
@@ -180,14 +181,102 @@ Result<std::shared_ptr<ArrayData>> GetTakeIndicesImpl(
+  const uint8_t* filter_selection = filter_values.buffers[1].data;
+  const bool filter_may_have_nulls = filter_values.MayHaveNulls();
+
+  // BinaryBitBlockCounter is not used here because a REE bitmap, if built
+  // correctly, is not going to have long continuous runs of 0s or 1s in the
+  // values array.
+
+  const ::arrow::ree_util::RunEndEncodedArraySpan<T> filter_span(filter);
+  auto it = filter_span.begin();
+  if (filter_may_have_nulls && null_selection == FilterOptions::EMIT_NULL) {
+    // Most complex case: the filter may have nulls and we don't drop them.
+    // The logic is ternary:
+    // - filter is null: emit null
+    // - filter is valid and true: emit index
+    // - filter is valid and false: don't emit anything
+
+    typename TypeTraits<RunEndType>::BuilderType builder(memory_pool);
+    for (; !it.is_end(filter_span); ++it) {
+      const int64_t position_with_offset = filter_values_offset + it.index_into_array();
+      const bool is_null = !bit_util::GetBit(filter_is_valid, position_with_offset);
+      if (is_null) {
+        RETURN_NOT_OK(builder.AppendNulls(it.run_length()));
+      } else {
+        const bool emit_run = bit_util::GetBit(filter_selection, position_with_offset);
+        if (emit_run) {
+          const int64_t run_end = it.run_end();
+          RETURN_NOT_OK(builder.Reserve(run_end - it.logical_position()));
+          for (int64_t position = it.logical_position(); position < run_end; position++) {
+            builder.UnsafeAppend(static_cast<T>(position));
+          }
+        }
+      }
+    }
+    std::shared_ptr<ArrayData> result;
+    RETURN_NOT_OK(builder.FinishInternal(&result));
+    return result;
+  }
+
+  // Other cases don't emit nulls and are therefore simpler.
+  TypedBufferBuilder<T> builder(memory_pool);
+
+  if (filter_may_have_nulls) {
+    DCHECK_EQ(null_selection, FilterOptions::DROP);
+    // The filter may have nulls, so we scan the validity bitmap and the filter
+    // data bitmap together.
+    for (; !it.is_end(filter_span); ++it) {
+      const int64_t position_with_offset = filter_values_offset + it.index_into_array();
+      const bool emit_run = bit_util::GetBit(filter_is_valid, position_with_offset) &&
+                            bit_util::GetBit(filter_selection, position_with_offset);
+      if (emit_run) {
+        const int64_t run_end = it.run_end();
+        RETURN_NOT_OK(builder.Reserve(run_end - it.logical_position()));
+        for (int64_t position = it.logical_position(); position < run_end; position++) {
+          builder.UnsafeAppend(static_cast<T>(position));
+        }
+      }
+    }
+  } else {
+    // The filter has no nulls, so we need only look for true values
+    for (; !it.is_end(filter_span); ++it) {
+      const int64_t position_with_offset = filter_values_offset + it.index_into_array();
+      const bool emit_run = bit_util::GetBit(filter_selection, position_with_offset);
```

Comment:
By the way, given the definition of REE, the run values should alternate between true and false. This might help make this loop faster (but might not). It can be left for a subsequent PR or GH issue.

Reply:
You're right. I was mistakingly assuming it was forbidden in the format.